sources/osgShadow/ShadowMapAtlas.js

```diff
@@ -350,55 +334,47 @@ ShadowMapAtlas.prototype = MACROUTILS.objectLibraryClass( MACROUTILS.objectInher
         return this._shadowMaps[ lightNum ];
     },
 
-    addLight: function ( light, settings, customMapSize ) {
+    addLight: function ( light ) {
 
-        if ( !light )
+        if ( !light || this._lights.indexOf( light ) !== -1 )
             return -1;
 
-        var lightNum = this._lights.indexOf( light );
-        if ( lightNum !== -1 )
-            return lightNum;
-
-        this._lights.push( light );
-        lightNum = this._lights.length;
-
-        if ( lightNum === 1 ) {
-            this._texture.setLightNumber( light.getLightNumber() );
-            this._texture.setLightUnit( light.getLightNumber() );
+        var lightCount = this._lights.length;
+        if ( lightCount === ( this._numShadowWidth * this._numShadowHeight ) ) {
+            notify.warn( 'can\'t allocate shadow for light ' + light.getLightNumber() + ' ShadowAtlas already full ' );
+            return undefined;
         }
 
-        var shadowMap = new ShadowMap( settings, this._texture );
+        this._lights.push( light );
+        this._shadowSettings.setLight( light );
+        var shadowMap = new ShadowMap( this._shadowSettings, this._texture );
         this._shadowMaps.push( shadowMap );
 
-        var mapSize = customMapSize !== undefined ? customMapSize : this._shadowMapSize;
-
-        var y = mapSize * ( lightNum % ( this._numShadowWidth ) );
-        var x = mapSize * ( Math.floor( lightNum / ( this._numShadowHeight ) ) );
+        var mapSize = this._shadowMapSize;
+        var y = mapSize * ( lightCount % ( this._numShadowWidth ) );
+        var x = mapSize * ( Math.floor( lightCount / ( this._numShadowHeight ) ) );
 
         shadowMap.setShadowedScene( this._shadowedScene );
         this._viewportDimension.push( vec4.fromValues( x, y, mapSize, mapSize ) );
 
         return shadowMap;
-    },
-
-    setLight: function ( light, lightNum ) {
-
-        if ( !light || this._lights.indexOf( light ) !== -1 )
-            return;
 
-        this._lights[ lightNum ] = light;
-        this.dirty();
     },
 
+
     /** initialize the ShadowedScene and local cached data structures.*/
     init: function () {
 
         if ( !this._shadowedScene ) return;
 
         this.initTexture();
+        var lightNumbers = [];
+        for ( var k = 0; k < this._lights.length; k++ ) {
+            lightNumbers.push( this._lights[ k ].getLightNumber() );
+        }
+        this._texture.setLightsIndexes( lightNumbers );
```

Comment:
renames to setLightNumberArray, it fit better what is it. lightIndex does not make sense, we should not use this term, too confusing